aws_quickstart/datadog_integration_api_call_v2.yaml

```diff
@@ -66,160 +66,194 @@ Resources:
-                  "LogicalResourceId": event['LogicalResourceId'],
-                  "Data": response_data
-              })
-              formatted_response = response_body.encode("utf-8")
-
-              LOGGER.info('ResponseURL: %s', event['ResponseURL'])
-              LOGGER.info('ResponseBody: %s', response_body)
-
-              opener = build_opener(HTTPHandler)
-              request = Request(event['ResponseURL'], data=formatted_response)
-              request.add_header('Content-Type', 'application/json; charset=utf-8')
-              request.add_header('Content-Length', len(formatted_response))
-              request.get_method = lambda: 'PUT'
-              response = opener.open(request)
-              LOGGER.info("Status code: %s", response.getcode())
-              LOGGER.info("Status message: %s", response.msg)
+                  # Call Datadog API and report response back to CloudFormation
+                  response = call_datadog_api(event, method)
+                  cfn_response_send_api_result(event, context, method, response)
+
+              except Exception as e:
+                  LOGGER.info("Failed - exception thrown during processing.")
+                  msg = {"Message": "Exception during processing: {}".format(e)}
+                  reason = json.dumps(msg)
+                  cfnresponse.send(
+                      event,
+                      context,
+                      "FAILED",
+                      responseData=msg,
+                      reason=reason,
+                  )
+
+
+          def cfn_response_send_api_result(event, context, method, response):
+              reason = None
+              json_response = ""
+              code = response.getcode()
+              data = response.read()
+              if data:
+                  json_response = json.loads(data)
+              if code == 200 or code == 204:
+                  LOGGER.info("Success - Datadog API call was successful.")
+                  response_status = "SUCCESS"
+                  msg = {"Message": "Datadog AWS Integration {} API request was successful.".format(method)}
+
+                  # return external ID for create and update
+                  if method == "POST" or method == "PATCH":
+                      external_id = json_response["data"]["attributes"]["aws_account"]["auth_config"]["external_id"]
+                      msg["ExternalId"] = external_id
+              else:
+                  LOGGER.info("Failed - Datadog API call failed.")
+                  response_status = "FAILED"
+                  msg = {
+                      "Message": "Datadog API returned error: {}".format(json_response)
+                  }
+                  reason = json.dumps(msg)  # surface error in CloudFormation console
+
+              cfnresponse.send(
+                  event,
+                  context,
+                  responseStatus=response_status,
+                  responseData=msg,
+                  reason=reason,
```

Comment:
This `reason` param allows us to display the actual API error directly in the Cloudformation UI (instead of requiring the user to navigate to their Cloudwatch logs stream for the lambda logs) 

![image](https://github.com/user-attachments/assets/145f79e5-a2c4-4996-abfc-6f689715796c)